src/query/functions/linear/histogram_quantile_test.go

```diff
@@ -0,0 +1,283 @@
+	validMeta3 := block.SeriesMeta{Tags: tags.Clone().SetBucket([]byte("10"))}
+	infMeta := block.SeriesMeta{Tags: tags.Clone().SetBucket([]byte("Inf"))}
+	validMetaMoreTags := block.SeriesMeta{Tags: tags.Clone().SetBucket([]byte("0.1")).AddTag(models.Tag{
+		Name:  []byte("qux"),
+		Value: []byte("qar"),
+	})}
+
+	metas := []block.SeriesMeta{
+		validMeta, noBucketMeta, invalidBucketMeta, validMeta2, validMetaMoreTags, validMeta3, infMeta,
+	}
+
+	actual := gatherSeriesToBuckets(metas)
+	expected := bucketedSeries{
+		`{bar="baz"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 0.1, idx: 0},
+				{upperBound: 0.1, idx: 3},
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(1), idx: 6},
+			},
+			tags: models.NewTags(1, tagOpts).AddTag(models.Tag{
+				Name:  []byte("bar"),
+				Value: []byte("baz"),
+			}),
+		},
+		`{bar="baz",qux="qar"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 0.1, idx: 4},
+			},
+			tags: models.NewTags(1, tagOpts).AddTag(models.Tag{
+				Name:  []byte("bar"),
+				Value: []byte("baz"),
+			}).AddTag(models.Tag{
+				Name:  []byte("qux"),
+				Value: []byte("qar"),
+			}),
+		},
+	}
+
+	assert.Equal(t, expected, actual)
+}
+
+func TestSanitizeBuckets(t *testing.T) {
+	bucketed := bucketedSeries{
+		`{bar="baz"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(1), idx: 6},
+				{upperBound: 1, idx: 0},
+				{upperBound: 2, idx: 3},
+			},
+		},
+		`{with="neginf"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(-1), idx: 6},
+				{upperBound: 1, idx: 0},
+				{upperBound: 2, idx: 3},
+			},
+		},
+		`{no="infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 0.1, idx: 4},
+				{upperBound: 0.2, idx: 14},
+				{upperBound: 0.3, idx: 114},
+			},
+		},
+		`{just="infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: math.Inf(1), idx: 4},
+			},
+		},
+		`{just="neg-infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: math.Inf(-1), idx: 4},
+			},
+		},
+	}
+
+	actual := bucketedSeries{
+		`{bar="baz"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 1, idx: 0},
+				{upperBound: 2, idx: 3},
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(1), idx: 6},
+			},
+		},
+	}
+
+	sanitizeBuckets(bucketed)
+	assert.Equal(t, actual, bucketed)
+}
+
+func TestBucketQuantile(t *testing.T) {
+	// single bucket returns nan
+	actual := bucketQuantile(0.5, []bucketValue{{upperBound: 1, value: 1}})
+	assert.True(t, math.IsNaN(actual))
+
+	// bucket with no infinity returns nan
+	actual = bucketQuantile(0.5, []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: 2, value: 2},
+	})
+	assert.True(t, math.IsNaN(actual))
+
+	actual = bucketQuantile(0.5, []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: math.Inf(1), value: 22},
+	})
+	assert.Equal(t, float64(1), actual)
```

Comment:
Add test with neg infinity and > 2 buckets?

Reply:
Good idea